src/lib/state/DY.Lib.State.Database.fst

```diff
@@ -0,0 +1,413 @@
+/// and searching to find records satisfying some property (analogous
+/// to database search queries).
+/// Its primary use case is when a protocol needs to maintain some global
+/// invariant across a particular kind of data for multiple sessions ---
+/// the prototypical example is ensuring that some session identifier is
+/// unique, and so no mix-up attacks arise spuriously from reuse of these
+/// identifiers.
+
+(*** Database state and invariants ***)
+
+/// Parameters for defining a database.
+/// The tag is used to identify a particular kind of database (since we may
+/// have multiple used by different levels of protocols in a system), and
+/// we also specify the row type, which needs to be able to be serialized
+/// as bytes.
+
+class db_types (row_t:Type0) = {
+  // TODO maybe these two tags can be merged into one input (and then derived tags created for the db and the row?)
+  db_tag: string;
+  row_tag: string;
+  ps_row_t: parser_serializer bytes row_t;
+}
+
+// TODO Probably need to rework to use reverse lists for this to make sense.
+#push-options "--ifuel 1"
+val list_growsP:
+  #a:Type0 ->
+  grows_pred:(a -> a -> prop) ->
+  l1:list a -> l2:list a ->
+  prop
+let rec list_growsP #a grows_pred l1 l2 =
+  match (l1, l2) with
+  | [], _ -> True
+  | h1::t1, [] -> False
+  | h1::t1, h2::t2 -> grows_pred h1 h2 /\ list_growsP grows_pred t1 t2
+#pop-options
+
+noeq type db_predicate {|crypto_invariants|} (row_t:Type0) {|db_t:db_types row_t|} = {
+  // TODO: Should the row_pred (and related) take a state ID?
+  // In principle, the user shouldn't be interacting with rows directly by state ID anyway
+  row_pred: trace -> principal -> row_t -> prop;
+  row_pred_later: tr1:trace -> tr2:trace -> prin:principal -> row:row_t -> Lemma
+    (requires row_pred tr1 prin row /\ tr1 <$ tr2)
+    (ensures row_pred tr2 prin row)
+  ;
+  row_pred_knowable: tr:trace -> prin:principal -> row:row_t -> Lemma
+    (requires row_pred tr prin row)
+    (ensures is_well_formed_prefix db_t.ps_row_t (is_knowable_by (principal_label prin) tr) row)
+  ;
+
+  // TODO: Should this also take two traces (trace when first state was set, trace now)?
+  row_update_pred: trace -> principal -> row_t -> row_t -> prop;
+
+  db_global_pred: trace -> principal -> state_id -> list row_t -> prop;
+  db_global_pred_empty: tr:trace -> prin:principal -> sess_id:state_id -> Lemma
+    (ensures db_global_pred tr prin sess_id [])
+  ;
+  db_global_pred_update: tr1:trace -> tr2:trace -> prin:principal -> sess_id:state_id -> db1:list row_t -> db2:list row_t -> Lemma
+    (requires db_global_pred tr1 prin sess_id db1 /\
+      tr1 <$ tr2 /\
+      list_growsP (row_update_pred tr2 prin) db1 db2
+    )
+    (ensures db_global_pred tr2 prin sess_id db2)
+  ;
+  // No knowable lemma needed for the overall database, because it holds by construction.
+}
+
+/// TODO give actual instances
+let ps_state_id : parser_serializer bytes state_id = admit()
```

Comment:
Naturally, these would need proper implementations, which I'm sure aren't hard, but also wasn't really the point of this draft.

Reply:
They do exist in `DY.Lib.Comparse.Parsers`